### PR TITLE
Exp: Use Muti-objective learning 

### PR DESCRIPTION
**Pull request by stevengogogo**

### CODE REVIEW

**Comment by stevengogogo** [comment]
Use $loss = loss_{drm} + 500 loss_{bc}$

```
python pinn_1d.py   --levels 1 --epochs 10000 --lr 1e-4 --activation gelu --sweeps 1 --hidden_dims 256 256 256 --high_freq 3 --loss_type 1  --bc_weight 1 --nx 2000  --gamma 0  --aggregator "Constant" 1. 500. --plot
```
```
Iteration    999/ 10000, DRM Loss: -3.5050e+01, Err 2-norm:  8.1668e-01, inf-norm: 1.1278e-01
```

**Comment by stevengogogo** [comment]
## Exp: Verify dual-cone optimization helps PINNs + soft BC loss

### Dual Cone 


```
python pinn_1d.py   --levels 1 --epochs 10000 --lr 1e-4 --activation gelu --sweeps 1 --hidden_dims 256 256 256 --high_freq 3 --loss_type 0  --bc_weight 1 --nx 2000  --gamma 0  --aggregator "UPGrad" --plot
```
```
Training Level 0
Level 0: TRAIN
Iteration    499/ 10000, PINN Loss: 1.1153e+01, Err 2-norm:  1.4233e+01, inf-norm: 1.4310e+00
Iteration    999/ 10000, PINN Loss: 1.0686e+00, Err 2-norm:  5.0536e-01, inf-norm: 4.5360e-02
Iteration   1499/ 10000, PINN Loss: 3.5229e-01, Err 2-norm:  2.2734e-01, inf-norm: 2.2083e-02
Iteration   1999/ 10000, PINN Loss: 1.1591e-01, Err 2-norm:  1.0089e-01, inf-norm: 1.1598e-02
Iteration   2499/ 10000, PINN Loss: 5.2839e-02, Err 2-norm:  4.8972e-02, inf-norm: 5.6659e-03
Iteration   2999/ 10000, PINN Loss: 2.9012e-02, Err 2-norm:  3.4169e-02, inf-norm: 3.8587e-03
Iteration   3499/ 10000, PINN Loss: 1.9204e-02, Err 2-norm:  2.0535e-02, inf-norm: 2.3413e-03
Iteration   3999/ 10000, PINN Loss: 1.3088e-02, Err 2-norm:  1.8215e-02, inf-norm: 2.0128e-03
Iteration   4499/ 10000, PINN Loss: 1.4971e-01, Err 2-norm:  5.1648e-02, inf-norm: 7.5520e-03
Iteration   4999/ 10000, PINN Loss: 7.1656e-03, Err 2-norm:  1.2028e-02, inf-norm: 1.9021e-03
Iteration   5499/ 10000, PINN Loss: 4.7366e-03, Err 2-norm:  7.8400e-03, inf-norm: 9.1186e-04
Iteration   5999/ 10000, PINN Loss: 3.6847e-03, Err 2-norm:  6.0205e-03, inf-norm: 7.1356e-04
Iteration   6499/ 10000, PINN Loss: 2.9949e-03, Err 2-norm:  4.2069e-03, inf-norm: 3.9807e-04
Iteration   6999/ 10000, PINN Loss: 2.4976e-03, Err 2-norm:  3.9032e-03, inf-norm: 4.7027e-04
Iteration   7499/ 10000, PINN Loss: 2.1062e-03, Err 2-norm:  3.2904e-03, inf-norm: 3.9752e-04
Iteration   7999/ 10000, PINN Loss: 1.7800e-03, Err 2-norm:  2.6893e-03, inf-norm: 3.2555e-04
Iteration   8499/ 10000, PINN Loss: 1.5204e-03, Err 2-norm:  2.2626e-03, inf-norm: 2.7375e-04
Iteration   8999/ 10000, PINN Loss: 1.3033e-03, Err 2-norm:  1.9027e-03, inf-norm: 2.2980e-04
Iteration   9499/ 10000, PINN Loss: 3.6962e-02, Err 2-norm:  5.8394e-02, inf-norm: 7.6404e-03
Iteration   9999/ 10000, PINN Loss: 9.7147e-04, Err 2-norm:  1.1256e-03, inf-norm: 1.2255e-04
```

image

snapshot of weights of gradients
```
Weights: tensor([0.5813, 1.0964], device='cuda:0')
Cosine similarity: 0.9370
Weights: tensor([0.5808, 1.0967], device='cuda:0')
Cosine similarity: 0.9372
Weights: tensor([0.5804, 1.0970], device='cuda:0')
Cosine similarity: 0.9374
```

## PINN with averaging loss

```
python pinn_1d.py   --levels 1 --epochs 10000 --lr 1e-4 --activation gelu --sweeps 1 --hidden_dims 256 256 256 --high_freq 3 --loss_type 0  --bc_weight 1 --nx 2000  --gamma 0 --plot
```

```
Iteration    499/ 10000, PINN Loss: 1.7111e+01, Err 2-norm:  3.9158e+00, inf-norm: 4.6693e-01
Iteration    999/ 10000, PINN Loss: 6.1949e-01, Err 2-norm:  1.1616e+00, inf-norm: 1.3581e-01
Iteration   1499/ 10000, PINN Loss: 1.9982e-01, Err 2-norm:  2.5898e-01, inf-norm: 3.0343e-02
Iteration   1999/ 10000, PINN Loss: 1.2159e-01, Err 2-norm:  1.2659e-01, inf-norm: 1.5461e-02
Iteration   2499/ 10000, PINN Loss: 7.9549e-02, Err 2-norm:  8.1205e-02, inf-norm: 9.7390e-03
Iteration   2999/ 10000, PINN Loss: 4.8445e-02, Err 2-norm:  4.6060e-02, inf-norm: 5.8364e-03
Iteration   3499/ 10000, PINN Loss: 3.3523e-02, Err 2-norm:  3.5849e-02, inf-norm: 4.3601e-03
Iteration   3999/ 10000, PINN Loss: 2.5072e-02, Err 2-norm:  2.8865e-02, inf-norm: 3.5170e-03
Iteration   4499/ 10000, PINN Loss: 1.9703e-02, Err 2-norm:  2.3881e-02, inf-norm: 2.9146e-03
Iteration   4999/ 10000, PINN Loss: 1.5842e-02, Err 2-norm:  2.0890e-02, inf-norm: 2.5257e-03
Iteration   5499/ 10000, PINN Loss: 1.3097e-02, Err 2-norm:  1.8221e-02, inf-norm: 2.1839e-03
Iteration   5999/ 10000, PINN Loss: 1.1010e-02, Err 2-norm:  1.6584e-02, inf-norm: 1.8699e-03
Iteration   6499/ 10000, PINN Loss: 9.4804e-03, Err 2-norm:  1.4153e-02, inf-norm: 1.6923e-03
Iteration   6999/ 10000, PINN Loss: 8.2994e-03, Err 2-norm:  1.2402e-02, inf-norm: 1.5738e-03
Iteration   7499/ 10000, PINN Loss: 2.5371e-02, Err 2-norm:  3.7283e-02, inf-norm: 3.5382e-03
Iteration   7999/ 10000, PINN Loss: 9.1092e-03, Err 2-norm:  1.5372e-02, inf-norm: 2.2640e-03
Iteration   8499/ 10000, PINN Loss: 5.9730e-03, Err 2-norm:  8.5018e-03, inf-norm: 1.0841e-03
Iteration   8999/ 10000, PINN Loss: 5.4161e-03, Err 2-norm:  8.3174e-03, inf-norm: 9.8834e-04
Iteration   9499/ 10000, PINN Loss: 4.9422e-03, Err 2-norm:  7.5278e-03, inf-norm: 8.9516e-04
Iteration   9999/ 10000, PINN Loss: 2.7889e-02, Err 2-norm:  4.5591e-02, inf-norm: 4.5103e-03
```

image
image

**Comment by stevengogogo** [comment]
@liruipeng  Please take a look. It's ready.

File: pinn/pinn_1d.py [hunk 1]
```diff
@@ -289,7 +293,7 @@ def super_loss(self, model, mesh, loss_func):
         x = mesh.x_train
         u = model.get_solution(x)
         loss = loss_func(u, mesh.u_ex)
-        return loss
+        return loss, [loss,]
```
Comment (stevengogogo) [inline pinn/pinn_1d.py:296]: major change
